clusterloader2/pkg/measurement/common/kube_state_metrics_measurement.go

```diff
@@ -0,0 +1,201 @@
+		selfPort:    8081,
+		metricsPort: 8080,
+	}
+}
+
+// Execute supports two actions:
+// - start - starts goroutine and queries /metrics every probeIntervalDefault interval,
+// it also collects initial latency metrics.
+// - gather - gathers latency metrics and creates a latency summary.
+func (m *ksmLatencyMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	if m.ctx == nil {
+		m.ctx, m.cancel = context.WithCancel(context.Background())
+	}
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+	client := config.ClusterFramework.GetClientSets().GetClient()
+	switch action {
+	case "start":
+		if m.isRunning {
+			klog.V(2).Infof("%s: measurement already running", m)
+			return nil, nil
+		}
+		m.stopCh = make(chan struct{})
+		// Start executing calls towards the kube-state-metrics /metrics endpoint
+		// every probeIntervalDefault until gather is called.
+		// probeIntervalDefault equals the scrapping interval we suggest.
+		// If we cannot get metrics for two minutes we are already going over
+		// the scrape interval so we should cancel.
+		err := m.startQuerying(m.ctx, client, probeIntervalDefault)
+		if err != nil {
+			return nil, err
+		}
+		// Retrieve initial latency when first call is done.
+		m.latency, err = m.retrieveKSMLatencyMetrics(m.ctx, client)
+		return nil, err
+	case "gather":
+		defer m.cancel()
+		return m.createKSMLatencySummary(m.ctx, client)
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+// createKSMLatencyReport gathers the latency one last time and creates the summary based on the Quantile from the sub histograms.
+// Afterwards it creates the Summary Report.
+func (m *ksmLatencyMeasurement) createKSMLatencySummary(ctx context.Context, client clientset.Interface) ([]measurement.Summary, error) {
+	if !m.isRunning {
+		klog.V(2).Infof("%s: measurement was not running", m)
+		return nil, nil
+	}
+	close(m.stopCh)
+	m.wg.Wait()
+	latestLatency, err := m.retrieveKSMLatencyMetrics(ctx, client)
+	if err != nil {
+		return nil, err
+	}
+	// We want to subtract the latest histogram from the first one we collect.
+	finalLatency := HistogramSub(latestLatency, m.latency)
+	// Pretty Print the report.
+	result := &measurementutil.LatencyMetric{}
+	err = SetQuantileFromHistogram(result, finalLatency)
+	if err != nil {
+		return nil, err
+	}
+	content, err := util.PrettyPrintJSON(result)
+	if err != nil {
+		return nil, err
+	}
+	// Create Summary.
+	return []measurement.Summary{measurement.CreateSummary(schedulerLatencyMetricName, "json", content)}, nil
+}
+
+// startQuerying queries /metrics endpoint of kube-state-metrics kube_ metrics every interval
+// and stops when stop is called.
+func (m *ksmLatencyMeasurement) startQuerying(ctx context.Context, client clientset.Interface, interval time.Duration) error {
+	m.isRunning = true
+	m.wg.Add(1)
+	var err error
+	go func() {
+		defer m.wg.Done()
+		for {
```

Comment:
Can you please move L147-L158 to a separate function (for easier readability)

Reply:
Do you mean the entire go func or just the inner for loop?

Reply:
Either would work :)
Moving the whole function is probably cleaner.

Reply:
@wojtek-t Thanks, I addressed my comments, please have another look.